junit5/src/main/java/org/jboss/weld/junit5/auto/ClassScanning.java

```diff
@@ -200,7 +200,7 @@ static void scanForRequiredBeanClass(List<Class<?>> testClasses, Weld weld, bool
         }
 
         for (Class<?> foundClass : foundClasses) {
-            if (hasBeanDefiningAnnotation(foundClass)) {
+            if (hasBeanDefiningAnnotation(foundClass) || testClasses.contains(foundClass)) {
```

Comment:
This is wrong, please revert it! The test class itself should not pass this condition. It will be registered as bean later on in the process via `TestInstanceInjectionExtension`. The trick is, that `ProcessAnnotatedType` will be fired for that class and we are adding the `@Singleton` annotation there. That is what will make it an actual bean so that later on we can perform more magic tricks with `ProcessInjectionTarget` on the instance provided from JUnit.

Reply:
It really does not work that way as I thought it would.

```java
package org.jboss.weld.junit5.auto.classscanning;

import static org.junit.jupiter.api.Assertions.assertFalse;

import javax.enterprise.inject.spi.CDI;

import org.jboss.weld.junit5.auto.EnableAutoWeld;
import org.junit.jupiter.api.Test;

@EnableAutoWeld
public class ClassScanningNotNecessarilyAddsTestClassTest {

    @Test
    void test() {
        assertFalse(CDI.current().select(ClassScanningNotNecessarilyAddsTestClassTest.class).isResolvable());
    }

}
```

All classes initially added to `classesToProcess` to start scanning with are always added to `foundClasses` without actually being scanned in `ClassScanning.scanForRequiredBeanClasses`.

Reply:
The code you posted will fail (e.g. `ClassScanningNotNecessarilyAddsTestClassTest` will become a bean) because of the CDI extension I described above.

> All classes initially added to classesToProcess to start scanning with are always added to foundClasses

If they pass [the condition](https://github.com/weld/weld-junit/blob/master/junit5/src/main/java/org/jboss/weld/junit5/auto/ClassScanning.java#L80-L85), then yes.

>  without actually being scanned in ClassScanning.scanForRequiredBeanClasses

What do you mean? It is within that method that this gets invoked. You always start with the test class and then based o what you find you are adding more classes to process.

Reply:
First I thought that class scanning would apply also to test classes, probably because there was that line I removed, see another discussion thread. But I agree that test classes, the outer-most one should be added always. And now it is added in `WeldJunit5AutoExtension` and not in `ClassScanning` and works well. To me it was not obvious that all test classes don't meet `if (hasBeanDefiningAnnotation(foundClass)) {` and still isn't and the test class would meet that condition if it had `@Singleton` already but gets that only afterwards. I changed it back.